File: 6kyu/xmas-tree-v1/src/XmasTree.ts
class XmasTree {

    readonly height: number;
    readonly tree: string[];

    private constructor(height: number, tree: string[]) {
        this.height = height;
        this.tree = tree;
    }

    public static fromHeight(height: number): XmasTree {
        const tree: string[] = [];
        for (let i = 1; i <= height; i++) {
            tree.push(this.generateTreeLevel(i, height));
        }
        return new XmasTree(height, tree);
    }

    public toString(): string {
        const strTree = this.tree.reduce((acc, curr, i, tree) => acc + (i !== tree.length - 1) ? "\n" : "", "");
        return strTree;
    }

    static generateTreeLevel(levelFromTop: number, height: number): string {
        const stars = this.generateStars(levelFromTop);
        const starsWithSpaces = this.addSpacesToStars(stars, height);
        return starsWithSpaces;
    }

    static addSpacesToStars(stars: string, height: number): string {
        const width = 2 * height - 1;
        const spacesPerSite = (width - stars.length) / 2;
        return " ".repeat(spacesPerSite) + stars;
    }

    static generateStars(count: number): string {
        if (count <= 1) {
            return "*";
        }
        return "**" + this.generateStars(count - 1);
    }

}

export default XmasTree;

File: Java/xmas-tree/xmas-tree-v2/src/main/java/de/klosebrothers/apprentice/kata/xmas_tree/XmasTree.java
package de.klosebrothers.apprentice.kata.xmas_tree;

import java.util.ArrayList;
import java.util.List;

public class XmasTree {

    private final int height;

    public static XmasTree fromHeight(int height) {
        return new XmasTree(height);
    }

    private XmasTree(int height) {
        this.height = height; w
    }

    @Override
    public String toString() {
        final List<String> tree = generateTreeFromHeight();
        String strTree = tree.stream().reduce()
        return tree.get(0);
    }

    private List<String> generateTreeFromHeight() {
        List<String> tree = new ArrayList<>();
        if (height == 0) {
            tree.add("");
            return tree;
        }
        tree.add("*");
        return tree;
    }

}
